### PR TITLE
Add schema migration init container [K8SSAND-625]

### LINKED ISSUE #56: Schema should be fully created prior to starting the Reaper pod
Schema migrations on startup of Reaper can take more time than the liveness probe would allow. This will cause the pods to be killed and restarted, which can happen in the middle of a migration. Some DDL statements are not idempotent (such as adding a column) and can put the pod in a loop as subsequent migration attempts will fail consistently.

Creating the whole schema upfront (including the schema migration table content) would ensure the Reaper pod starts quickly on the first run and avoid that behavior.

This could be done through a job or an init container.

┆Issue is synchronized with this [Jira Bug](https://k8ssandra.atlassian.net/browse/K8SSAND-625) by [Unito](https://www.unito.io)
┆Fix Versions: k8ssandra-1.3.0

### PR DESCRIPTION
**Pull request by adejanovski**
This PR requires Reaper [PR 1096](https://github.com/thelastpickle/cassandra-reaper/pull/1096) to be merged and released first.

Fixes #56 

This change adds an init container that will start Reaper in schema migration only mode, allowing for it to pass without probes that could trigger a restart mid-migration.

### CODE REVIEW

File: pkg/reconcile/reconcilers.go
Comment (burmanm) [inline pkg/reconcile/reconcilers.go:496]: Shouldn't this be len(envVars) + 1 ? Instead of letting append later on to grow the size again.

**Comment by adejanovski** [comment]
e2e tests are fixed and things work as expected.
We'll have to wait until a new Docker tag is released for Reaper before merging the PR, but it would be good to get an approval ASAP so that we can get this into 1.3.

File: test/config/reaper/bases/reaper.yaml
Comment (burmanm) [inline test/config/reaper/bases/reaper.yaml:6]: Should probably be in k8ssandra repo.

Comment (jsanda) [reply test/config/reaper/bases/reaper.yaml:6]: Because of Reaper and Medusa being used outside of k8ssandra I believe the decision was made to keep them
in the TLP org. On that basis I don't see a need to publish the image under k8ssandra.

Comment (burmanm) [reply test/config/reaper/bases/reaper.yaml:6]: But that's not in the TLP org either.

Comment (adejanovski) [reply test/config/reaper/bases/reaper.yaml:6]: I didn't mean to commit the PR like this, as said in my comment :) 

> We'll have to wait until a new Docker tag is released for Reaper before merging the PR, but it would be good to get an approval ASAP so that we can get this into 1.3.

I have mixed feelings about merging the PR on a tag that's not an actual release, but we'll probably have to wait a little bit more until we get a new release of Reaper. I may just create a tag with a name that's easy enough to spot in the tlp org and update this PR so it can be merged.

Comment (adejanovski) [reply test/config/reaper/bases/reaper.yaml]: I've moved to a tag on the tlp repo now. I'll try to remember to upgrade this when Reaper 2.3.0 gets released.

Comment (jdonenine) [reply test/config/reaper/bases/reaper.yaml]: So are we good to put this into 1.3.0 before Reaper 2.3.0 is out?

I just talked to Erik earlier and we're looking at a few more days for the incremental repair work, so if we were waiting for that before releasing Reaper we may want to just go ahead and put out a patch.

Comment (adejanovski) [reply test/config/reaper/bases/reaper.yaml]: yes, we can go ahead and merge it as is. I'd like us to get the incremental repair stuff in before releasing 2.3.0.

Comment (jsanda) [reply test/config/reaper/bases/reaper.yaml]: Assuming Reaper 2.3.0 will be released before k8ssandra 1.3.0 then we can open a follow up PR to bump the default version. I am in favor of merging and updating in k8ssandra to give the changes some soak time before releasing 1.3.0.

**Review by jsanda — APPROVED** [review]

File: test/config/reaper/bases/reaper.yaml
Comment (jdonenine) [reply test/config/reaper/bases/reaper.yaml]: The trick here is that I don't think we're going to get reaper-2.3.0 out before k8ssandra-1.3.0.  So if we want to include this in k8ssandra-1.3.0 we'll need to use some patch version of reaper that we could get out sooner.

@adejanovski what do you think about releasing a v2.2.6 of Reaper that would pickup your work in https://github.com/thelastpickle/cassandra-reaper/pull/1096 and use that?

We could hopefully then get reaper-2.3.0 out post k8ssandra-1.3.0 and use that in k8ssandra-1.4.0 to pickup the incremental repair support?

Comment (adejanovski) [reply test/config/reaper/bases/reaper.yaml]: > @adejanovski what do you think about releasing a v2.2.6 of Reaper that would pickup your work in thelastpickle/cassandra-reaper#1096 and use that?

It's doable but slightly breaks the contract that patch releases are for fixes only.

Since we have no solid ETA for incremental repair improvements, I guess we could release 2.3.0 right now and delay other planned features to 2.4.0.
wdyt @jdonenine ?

Comment (jdonenine) [reply test/config/reaper/bases/reaper.yaml]: That works for me @adejanovski 

I'd vote we go ahead and release 2.3.0 and we'll come back with 2.4.0 for incremental repair when it's ready.

Comment (jdonenine) [reply test/config/reaper/bases/reaper.yaml]: For the record, so you don't think I've gone crazy, I was thinking a patch release because what we're generally trying to fix is a bug:  https://github.com/k8ssandra/reaper-operator/issues/56

But you're totally right that what we added to reaper itself, to address the bug over here in K8ssandra, isn't a bug fix.  So I think you've got the right suggestion with bump to 2.4.0.

**Pull request merged**
